File: lang/java/reef-bridge-java/src/main/java/org/apache/reef/bridge/NetworkTransport.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.reef.bridge;

import org.apache.avro.specific.SpecificRecord;
import org.apache.reef.tang.InjectionFuture;
import org.apache.reef.wake.EventHandler;
import org.apache.reef.wake.MultiObserver;
import org.apache.reef.wake.avro.ProtocolSerializer;
import org.apache.reef.wake.remote.RemoteMessage;
import org.apache.reef.wake.remote.address.LocalAddressProvider;
import org.apache.reef.wake.remote.impl.ByteCodec;
import org.apache.reef.wake.remote.impl.SocketRemoteIdentifier;
import org.apache.reef.wake.remote.RemoteIdentifier;
import org.apache.reef.wake.remote.RemoteManager;
import org.apache.reef.wake.remote.RemoteManagerFactory;

import javax.inject.Inject;
import java.net.InetSocketAddress;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * The CLR Bridge Network class aggregates a RemoteManager and
 * Protocol Serializer to provide a simple send/receive interface
 * between the Java and CLR sides of the bridge.
 */
public final class NetworkTransport {

  private static final Logger LOG = Logger.getLogger(NetworkTransport.class.getName());

  /** Remote manager to handle java-C# bridge communication. */
  private final RemoteManager remoteManager;

  private final ProtocolSerializer serializer;
  private final InetSocketAddress inetSocketAddress;
  private final InjectionFuture<MultiObserver> messageObserver;

  private EventHandler<byte[]> sender;

  /**
   * Sends and receives messages between the java bridge and C# bridge.
   * @param observer A multiobserver instance that will receive all incoming messages.
   */
  @Inject
  private NetworkTransport(
      final RemoteManagerFactory remoteManagerFactory,
      final LocalAddressProvider localAddressProvider,
      final ProtocolSerializer serializer,
      final InjectionFuture<MultiObserver> observer) {

    LOG.log(Level.FINE, "Java bridge network initializing");

    this.serializer = serializer;
    this.messageObserver = observer;

    this.remoteManager = remoteManagerFactory.getInstance(
        "JavaBridgeNetwork", localAddressProvider.getLocalAddress(), 0, new ByteCodec());

    // Get our address and port number.
    final RemoteIdentifier remoteIdentifier = this.remoteManager.getMyIdentifier();
    if (remoteIdentifier instanceof SocketRemoteIdentifier) {
      final SocketRemoteIdentifier socketIdentifier = (SocketRemoteIdentifier)remoteIdentifier;
      this.inetSocketAddress = socketIdentifier.getSocketAddress();
    } else {
      throw new RuntimeException("Identifier is not a SocketRemoteIdentifier: " + remoteIdentifier);
    }

    // Register as the message handler for any incoming messages.
    this.remoteManager.registerHandler(byte[].class, new LocalObserver());
  }

  /**
   * Sends a message to the C# side of the bridge.
   * @param message An Avro message class derived from SpecificRecord.
   * @throws RuntimeException if invoked before initialization is complete.
   */
  public void send(final long identifier, final SpecificRecord message) {
    if (sender != null) {
      sender.onNext(serializer.write(message, identifier));
    } else {
      final String msgClassName = message.getClass().getCanonicalName();
      LOG.log(Level.SEVERE, "Attempt to send message [{0}] before network is initialized", msgClassName);
      throw new RuntimeException("NetworkTransport not initialized: failed to send " + msgClassName);
    }
  }

  /**
   * Provides the IP address and port of the java bridge network.
   * @return A InetSockerAddress that contains the ip and port of the bridge network.
   */
  public InetSocketAddress getAddress() {
    return inetSocketAddress;
  }

  /**
   * Processes messages from the network remote manager.
   */
  private final class LocalObserver implements EventHandler<RemoteMessage<byte[]>> {

    /**
     * Deserialize and direct incoming messages to the registered MuiltiObserver event handler.
     * @param message A RemoteMessage<byte[]> object which will be deserialized.
     */
    @Override
    public void onNext(final RemoteMessage<byte[]> message) {
      LOG.log(Level.FINEST, "Received remote message: {0}", message);

      if (sender == null) {
        // Instantiate a network connection to the C# side of the bridge.
        // THERE COULD BE  A SECURITY ISSUE HERE WHERE SOMEONE SPOOFS THE
        // C# BRIDGE, WE RECEIVE IT FIRST, AND CONNECT TO THE SPOOFER,
        // THOUGH THE TIME WINDOW IS VERY SMALL.
        final RemoteIdentifier remoteIdentifier = message.getIdentifier();
        LOG.log(Level.FINE, "Connecting to: {0}", remoteIdentifier);
        sender = remoteManager.getHandler(remoteIdentifier, byte[].class);
      }

      // Deserialize the message and invoke the appropriate processing method.
      serializer.read(message.getMessage(), messageObserver.get());
    }
  }
}

File: lang/cs/Org.Apache.REEF.Bridge.CLR/NetworkTransport.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System;
using System.Collections.Concurrent;
using System.IO;
using System.Net;
using org.apache.reef.bridge.message;
using Org.Apache.REEF.Common.Files;
using Org.Apache.REEF.Tang.Annotations;
using Org.Apache.REEF.Utilities.Logging;
using Org.Apache.REEF.Wake.Avro;
using Org.Apache.REEF.Wake.Remote;
using Org.Apache.REEF.Wake.Remote.Impl;

namespace Org.Apache.REEF.Bridge
{
    /// <summary>
    /// The CLR Bridge Network class agregates a RemoteManager and
    /// Protocol Serializer to provide a simple send/receive interface
    /// between the CLR and Java bridges.
    /// </summary>
    public sealed class NetworkTransport
    {
        private static readonly Logger Logger = Logger.GetLogger(typeof(NetworkTransport));

        private readonly BlockingCollection<byte[]> queue = new BlockingCollection<byte[]>();

        private readonly ProtocolSerializer _serializer;
        private readonly IRemoteManager<byte[]> _remoteManager;
        private readonly IObserver<byte[]> _remoteObserver;
        private readonly REEFFileNames _fileNames;

        /// <summary>
        /// Construct a network stack using the wake remote manager.
        /// </summary>
        /// <param name="localAddressProvider">An address provider used to obtain
        /// a local IP address on an open port.</param>
        /// <param name="serializer">Serializer/deserializer of the bridge messages.</param>
        /// <param name="localObserver">Handler of the incoming bridge messages.</param>
        [Inject]
        public NetworkTransport(
            ILocalAddressProvider localAddressProvider,
            ProtocolSerializer serializer,
            LocalObserver localObserver,
            IRemoteManagerFactory remoteManagerFactory,
            REEFFileNames fileNames)
        {
            _serializer = serializer;
            _fileNames = fileNames;

            // Instantiate the remote manager.
            _remoteManager = remoteManagerFactory.GetInstance(localAddressProvider.LocalAddress, new ByteCodec());

            // Listen to the java bridge on the local end point.
            _remoteManager.RegisterObserver(localObserver);
            Logger.Log(Level.Info, "Local observer listening to java bridge on: [{0}]", _remoteManager.LocalEndpoint);

            // Instantiate a remote observer to send messages to the java bridge.
            IPEndPoint javaIpEndPoint = GetJavaBridgeEndpoint();
            Logger.Log(Level.Info, "Connecting to java bridge on: [{0}]", javaIpEndPoint);
            _remoteObserver = _remoteManager.GetRemoteObserver(javaIpEndPoint);

            // Negotiate the protocol.
            Send(0, new BridgeProtocol(100));
        }

        /// <summary>
        /// Send a message to the java side of the bridge.
        /// </summary>
        /// <param name="identifier">A long value that which is the unique sequence identifier of the message.</param>
        /// <param name="message">An object reference to a message in the org.apache.reef.bridge.message package.</param>
        public void Send(long identifier, object message)
        {
            Logger.Log(Level.Verbose, "Sending message: {0}", message);
            _remoteObserver.OnNext(_serializer.Write(message, identifier));
        }

        /// <summary>
        /// Retrieves the address of the java bridge.
        /// </summary>
        /// <returns>IP address and port of the Java bridge.</returns>
        private IPEndPoint GetJavaBridgeEndpoint()
        {
            using (FileStream stream = File.Open(_fileNames.GetDriverJavaBridgeEndpoint(), FileMode.Open))
            {
                using (StreamReader reader = new StreamReader(stream))
                {
                    string javaBridgeAddress = reader.ReadToEnd();
                    Logger.Log(Level.Info, "Java bridge address: {0}", javaBridgeAddress);

                    string[] javaAddressStrs = javaBridgeAddress.Split(':');
                    IPAddress javaBridgeIpAddress = IPAddress.Parse(javaAddressStrs[0]);
                    int port = int.Parse(javaAddressStrs[1]);
                    return new IPEndPoint(javaBridgeIpAddress, port);
                }
            }
        }
    }
}
